MobileCenter/MobileCenter/Internals/Channel/MSChannelDefault.m

```diff
@@ -392,4 +399,58 @@ - (void)notifyFailureBeforeSendingForItem:(id<MSLog>)item withError:(NSError *)e
   }
 }
 
+#pragma mark – Observers
+
+- (void)addObservers {
+
+// There is no need to do trigger sending on macOS because we can just continue to execute tasks there.
+#if !TARGET_OS_OSX
+  if (!MS_IS_APP_EXTENSION) {
+    __weak typeof(self) weakSelf = self;
+    if (self.appDidEnterBackgroundObserver == nil) {
+      void (^notificationBlock)(NSNotification *note) = ^(NSNotification __unused *note) {
+        typeof(self) strongSelf = weakSelf;
+        if (self.timerSource != nil) {
+          [strongSelf flushQueue];
+
+          /**
+           * From the documentation for applicationDidEnterBackground:
+           * "It's likely any background tasks you start in applicationDidEnterBackground: will not run until after that
+           * method exits, you should request additional background execution time before starting those tasks.
+           * In other words, first call beginBackgroundTaskWithExpirationHandler: and then run the task on a
+           * dispatch queue or secondary thread.
+           */
+          UIApplication *sharedApplication = [MSUtility sharedApplication];
+
+          // Checking if sharedApplication is != nil as it can be nil for extensions.
+          if (sharedApplication) {
+            __block UIBackgroundTaskIdentifier _backgroundTask =
+                [sharedApplication beginBackgroundTaskWithExpirationHandler:^{
+                  [sharedApplication endBackgroundTask:_backgroundTask];
```

Comment:
One other thing, looks like those tasks don't get ended automatically when the application gets back into the foreground. This means they will stack with the next bg tasks and they will all be ended on the next expiration. Knowing that there are as many bg tasks as channels we are going to stack quite many. Maybe we should go clean and end those tasks in the didfinishlaunchingwithoptions. Or is it me being paranoid?

Reply:
Hm...I don't think that you are too paranoid. I'll come up with something.